### PR TITLE
Updates to yaml allowing decoding into structs and maintaining field positions and comments.

### PR DESCRIPTION
**Pull request by taustgen-wework**
Creates a new StructMeta type that can be placed on structs to store yaml metadata such as comments and field ordering in order to facilitate update in place.

Also makes the indent level of block sequences configurable.

### CODE REVIEW

File: encode.go [hunk 1]
```diff
@@ -128,13 +128,13 @@ func (e *encoder) marshal(tag string, in reflect.Value) {
 		e.nodev(in.Addr())
 		return
 	case time.Time:
-		e.timev(tag, in)
+		e.timev(tag, in, head, line, foot)
```
Comment (hardboiled) [inline encode.go]: would making head/line/foot a datastructure so it only needs to be passed as one param be an improvement?

File: encode.go [hunk 2]
```diff
@@ -211,13 +211,50 @@ func (e *encoder) fieldByIndex(v reflect.Value, index []int) (field reflect.Valu
...
-	e.mappingv(tag, func() {
+
+	fieldsIndex := sinfo.FieldsList
+
+	headComments, lineComments, footComments := makeEmptyComments(len(fieldsIndex))
+
+	if fIndex := getYamlMeta(in, fieldsIndex); fIndex.IsValid() {
+		meta := fIndex.Elem().Interface().(StructMeta)
+		fieldsIndex = meta.GetFieldsIndex()
+		headComments, lineComments, footComments = meta.GetComments()
```
Comment (we4tech) [inline encode.go]: 👍

File: encode.go [hunk 3]
```diff
@@ -102,13 +102,13 @@ func (e *encoder) marshalDoc(tag string, in reflect.Value) {
 	} else {
 		yaml_document_start_event_initialize(&e.event, nil, nil, true)
 		e.emit()
-		e.marshal(tag, in)
+		e.marshal(tag, in, nil, nil, nil)
```
Comment (hardboiled) [inline encode.go]: Maybe we can have a marshallWithMetadata function to that marshall wraps so that we can reduce the number of line changes?

```golang
func (e *encoder) marshal(tag string, in reflect.Value) {
  e.marsallWithMetadata(tag, in, nil, nil, nil)
}
```

in general is there a way to clearly separate the parts of the code that use metadata and the parts that don't?

File: encode.go [hunk 4]
```diff
@@ -211,13 +211,50 @@ func (e *encoder) fieldByIndex(v reflect.Value, index []int) (field reflect.Valu
...
+			e.flow = info.Flow
+			e.marshal("", value, headComments[i][1], lineComments[i][1], footComments[i][1])
+			processed[info.Id] = true
+		}
+
 		for _, info := range sinfo.FieldsList {
+			if _, done := processed[info.Id]; done {
+				continue
+			}
+			if info.Key == "yaml_meta" {
```
Comment (we4tech) [inline encode.go]: can we use a constant to declare `"yaml_meta"`? So that we can cross-reference through the constant

File: encode_test.go [hunk 5]
```diff
@@ -656,6 +656,17 @@ func (s *S) TestSetIndent(c *C) {
 	c.Assert(buf.String(), Equals, "a:\n        b:\n                c: d\n")
 }
 
+func (s *S) TestSequenceUndent(c *C) {
+	var buf bytes.Buffer
+	enc := yaml.NewEncoder(&buf)
+	enc.SetSequenceUndent(true)
```
Comment (we4tech) [inline encode_test.go:662]: What is sequence undent?

File: encode_test.go [hunk 6]
```diff
@@ -656,6 +656,17 @@ func (s *S) TestSetIndent(c *C) {
...
+	err := enc.Encode(map[string]interface{}{"a": []interface{}{"b", "c", "d"}})
+	c.Assert(err, Equals, nil)
+	err = enc.Close()
```
Comment (we4tech) [inline encode_test.go:665]: Can we use defer? Also, do we need a test for the `close` func?

File: structmeta_internal_test.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+	C    string
+}
+
+var structMetaTests = []string{
+	"a: a\nb:\n    c: c\n    d: d\n",
+	"b: #foo\n    c: c #bar\n    d: d #baz\na: a #end\n",
+	"a: ant #ant\n# a\nb: #beeline\n    c: cockroach #cockroach\n    #c\n    #d\n    d: dragonfly\n    #dragonfly\n",
+}
+
+func (s *S) TestStructMeta(c *C) {
```
Comment (we4tech) [inline structmeta_internal_test.go:33]: 👍

File: yaml.go [hunk 8]
```diff
@@ -278,6 +278,11 @@ func (e *Encoder) SetIndent(spaces int) {
 	e.encoder.indent = spaces
 }
 
+//
```
Comment (we4tech) [inline yaml.go]: Can you comment on this function?

File: yaml.go [hunk 9]
```diff
@@ -621,6 +625,9 @@ func getStructInfo(st reflect.Type) (*structInfo, error) {
 		}
 
 		if tag != "" {
+			if tag == "yaml_meta" && field.Type.String() != "yaml.StructMeta" {
```
Comment (we4tech) [inline yaml.go]: Do we need nesting?

File: encode.go [hunk 10]
```diff
@@ -244,43 +283,49 @@ func (e *encoder) structv(tag string, in reflect.Value) {
...
-func (e *encoder) mappingv(tag string, f func()) {
+func (e *encoder) mappingv(tag string, head, line, foot []byte, f func()) {
 	implicit := tag == ""
 	style := yaml_BLOCK_MAPPING_STYLE
 	if e.flow {
 		e.flow = false
 		style = yaml_FLOW_MAPPING_STYLE
 	}
 	yaml_mapping_start_event_initialize(&e.event, nil, []byte(tag), implicit, style)
+	e.event.head_comment = head
```
Comment (hardboiled) [inline encode.go]: Should we add a comment here that explains that the placement of assignments for `e.event` and the emits have to be done in this particular order?

File: decode.go [hunk 11]
```diff
@@ -851,7 +851,33 @@ func isStringMap(n *Node) bool {
 	return true
 }
 
+// StuctMeta can be added to a struct and  holds the comments and field ordering for that struct
+type StructMeta interface {
+	GetFieldsIndex() []fieldInfo
```
Comment (we4tech) [inline decode.go:859]: Can you comment the func?

File: scannerc.go [hunk 12]
```diff
@@ -2847,7 +2847,7 @@ func yaml_parser_scan_line_comment(parser *yaml_parser_t, token_mark yaml_mark_t
 			continue
 		}
 		if parser.buffer[parser.buffer_pos+peek] == '#' {
-			seen := parser.mark.index+peek
+			seen := parser.mark.index + peek
```
Comment (hardboiled) [inline scannerc.go:2850]: Are these changes intended? Should we revert this file?

File: go.mod [hunk 13]
```diff
@@ -1,5 +1,5 @@
-module "gopkg.in/yaml.v3"
+module gopkg.in/yaml.v3
 
-require (
-	"gopkg.in/check.v1" v0.0.0-20161208181325-20d25e280405
-)
+go 1.15
```
Comment (hardboiled) [inline go.mod]: Are we going to force the maintainers to update to go 1.15?

**Review by hardboiled — CHANGES_REQUESTED** [review]

File: encode_test.go (hunk 5 above)
Comment (taustgen-wework) [reply encode_test.go:662]: Between yaml v2 and v3 the way sequences is presented changed from 

```
list:
- li1
- li2
```
to
```
list:
  - li1
  - li2
```


Since the default needs to be false in go, I cant simply use `SetSequenceIndent` with a default of true. So `SetSequenceRemoveIndent`, `SetSequenceDedent` and `SetSequenceUnindent` were my other considered names

File: yaml.go (hunk 9 above)
Comment (taustgen-wework) [reply yaml.go]: Good point we actually dont.

tag == "yaml_meta"  implies the other is true so we can move it out of the `tag != ""`

File: decode.go (hunk 11 above)
Comment (taustgen-wework) [reply decode.go:859]: 👍

File: encode.go (hunk 3 above)
Comment (taustgen-wework) [reply encode.go]: Unfortunately not, its all recursive calls through `marshal` for the various arrays/structs/maps until it hits the bottom and uses `stringv` `boolv` `intv` `uintv` `floatv` or `nilv` to convert the primitve to string and call `emitScalar`.

To use metaData I need comments to exist on the recursive function, so that they can get passed down to the correct primitive call `stringv` `intv` etc. So it would end up duplicating essentially the entire function with 2 seperate recursive paths where the first can cutover to the second at any point.

File: encode.go (hunk 1 above)
Comment (taustgen-wework) [reply encode.go]: It is and done

File: encode.go (hunk 4 above)
Comment (taustgen-wework) [reply encode.go]: 👍

File: go.mod (hunk 13 above)
Comment (taustgen-wework) [reply go.mod]: image

File: scannerc.go (hunk 12 above)
Comment (taustgen-wework) [reply scannerc.go:2850]: I'm torn, these aren't intended. It's a result of having changes in this file and then removing them, the autoformatter still runs on save. 

I'm usually opposed to unintended changes, but it also feels wrong to intentional rebreak style. So I would argue its a beneficial change to fix style even when I'm not otherwise touching the filee

File: encode_test.go (hunk 6 above)
Comment (taustgen-wework) [reply encode_test.go:665]: I don't think I can use defer here as `enc.Close()` flushes the buffer and I can't assert the change until after that is done. Defer would mean it runs when the surrending function returns.


As for a seperate test for that, I'm not sure we gain additional confidence as it is implicitly tested in both my new test and all the old encoder settings tests.

File: yaml.go (hunk 8 above)
Comment (taustgen-wework) [reply yaml.go]: 👍

File: encode.go (hunk 10 above)
Comment (taustgen-wework) [reply encode.go]: Added Comment!

**Review by we4tech — APPROVED** [review]

File: go.mod (hunk 13 above)
Comment (taustgen-wework) [reply go.mod]: and it came back grr. Deleted again

**Comment by taustgen-wework** [comment]
go version comes back every time I run `go test ./...` and I didnt notice. Extra commit removes `go 1.15` directive
